Update to new enum extension system (#103)
neoforged/Documentation: docs/advanced/extensibleenums.md

All enums that can have new entries implement the `IExtensibleEnum` interface. This interface acts as a marker to allow the `RuntimeEnumExtender` launch plugin service to know what enums should be transformed.

:::warning
You should **not** be implementing this interface on your own enums. Use maps or registries instead depending on your usecase.  
Enums which are not patched to implement the interface cannot have the interface added to them via mixins or coremods due to the order the transformers run in.
:::

### Creating an Enum Entry

> To create new enum entries, a JSON file needs to be created and referenced in the `neoforge.mods.toml` with the `enumExtender` entry of a `[[mods]]` block. The specified path must be relative to the `resources` directory.

Reviewer: Example with a recommended file location here please.

The definition of the entry consists of the target enum's class name, the new field's name (must be prefixed with the mod ID), the descriptor of the constructor to use for constructing the entry and the parameters to be passed to said constructor.

> ```json5

Reviewer: Reformat JSON to 2 spaces.

            "constructor": "(ILjava/lang/String;Ljava/util/function/UnaryOperator;)V",
            // The parameters to be used, provided as a reference to an EnumProxy<Rarity> field in the given class
            "parameters": {
                "class": "example/examplemod/MyEnumParams",
                "field": "CUSTOM_RARITY_ENUM_PROXY"
            }
        },
        {
            "enum": "net/minecraft/world/damagesource/DamageEffects",
            "name": "EXAMPLEMOD_TEST",
            "constructor": "(Ljava/lang/String;Ljava/util/function/Supplier;)V",
            // The parameters to be used, provided as a reference to a method in the given class
            "parameters": {
                "class": "example/examplemod/MyEnumParams",
                "field": "getTestDamageEffectsParameter"
            }
        },
        {
            "enum": "net/minecraft/world/item/ItemDisplayContext",
            "name": "EXAMPLEMOD_STANDING",
            "constructor": "(ILjava/lang/String;Ljava/lang/String;)V",
            // Constant parameters provided directly.
            // This enum has an ID parameter which enforces providing -1 in the case of constant parameters

>             // This enum has a name which enforces prefixing with the mod ID

Reviewer: I would argue this to be unnecessary context here.

            "parameters": [ -1, "examplemod:standing", null ]
        }
    ]
}
```

```java
public class MyEnumParams {
    public static final EnumProxy<Rarity> CUSTOM_RARITY_ENUM_PROXY = new EnumProxy<>(
            Rarity.class, -1, "examplemod:custom", (UnaryOperator<Style>) style -> style.withItalic(true)
    );
    
    public static Object getTestDamageEffectsParameter(int idx, Class<?> type) {
        return type.cast(switch (idx) {
            case 0 -> "examplemod:test";
            case 1 -> (Supplier<SoundEvent>) () -> SoundEvents.DONKEY_ANGRY;
            default -> throw new IllegalArgumentException("Unexpected parameter index: " + idx);
        });
    }
}
```

#### Constructor

> The constructor must be specified as a method descriptor and must only contain the parameters visible in the source code, omitting the hidden constant name and ordinal parameters

Reviewer: ```suggestion
The constructor must be specified as a method descriptor and must only contain the parameters visible in the source code, omitting the hidden constant name and ordinal parameters.
```

#### Parameters

The parameters can be specified in three ways with limitations depending on the parameter types:

> - Inline in the JSON file as an array of constants (only allowed for primitive values, Strings and for passing null to any reference type)

Reviewer: ```suggestion
- Inline the parameter values in the JSON file as an array of constants. Only allowed for primitive values, strings and for passing null to any reference type.
```

Author: I'm not a fan of that wording in a listing. I've reworded the points a bit, let me know if that's better.

> - As a reference to a field of type `EnumProxy<TheEnum>` in a class from the mod

Reviewer: ```suggestion
- Specify a reference to a field of type `EnumProxy<TheEnum>` in a class from the mod (see `EnumProxy` example above).
```

> - As a reference to a method taking a parameter index and a `Class` object as parameters and returning `Object`
>   - The `Class` object is the type expected for the given parameter and should be used to cast (`Class#cast()`) the return value in order to keep `ClassCastException`s in mod code

Reviewer: ```suggestion
- Specify a reference to a method returning `Object`, where the return value is the parameter value to use. The method must have exactly two parameters of type `int` (index of the parameter) and `Class<?>` (expected type of the parameter) that should be used to determine the parameter value.
```

> :::warning
> The fields and/or methods used as parameter sources should be in a separate class to avoid unintentionally loading mod classes too early
> :::

Reviewer: ```suggestion
:::tip
The fields and/or methods used as sources for parameter values should be in a separate class to avoid unintentionally loading mod classes too early.
:::
```

Author: I disagree with making this a tip, it's not merely a nice-to-have.

Certain parameters have additional rules:

> - If the parameter is an int ID parameter related to a `@IndexedEnum` annotation on the enum, then it is ignored and replaced by the entry's ordinal. If said parameter is specified inline in the JSON, then it must be specified as `-1`

Reviewer: ```suggestion
- If the parameter is an int ID parameter related to a `@IndexedEnum` annotation on the enum, then it is ignored and replaced by the entry's ordinal. If said parameter is specified inline in the JSON, then it must be specified as `-1`.
```

> - If the parameter is a String name parameter related to a `@NamedEnum` annotation on the enum, then it must be prefixed by the mod ID in the `namespace:path` format known from `ResourceLocation`s

Reviewer: ```suggestion
- If the parameter is a String name parameter related to a `@NamedEnum` annotation on the enum, then it must be prefixed by the mod ID in the `namespace:path` format known from `ResourceLocation`s.
```
Maybe also add a link to the resource locations article.

> #### Retrieving the generated constant

Reviewer: ```suggestion
#### Retrieving the Generated Constant
```

> The generated enum constant can be retrieved via `TheEnum.valueOf(String)`. If a field reference is used to provide the parameters, then the constant can also be retrieved from the `EnumProxy` object via `EnumProxy#getValue()`

Reviewer: ```suggestion
The generated enum constant can be retrieved via `TheEnum.valueOf(String)`. If a field reference is used to provide the parameters, then the constant can also be retrieved from the `EnumProxy` object via `EnumProxy#getValue()`.
```

## Contributing to NeoForge

To add a new extensible enum to NeoForge, there are at least two required things to do:

> - First, have the enum implement `IExtensibleEnum` to mark that this enum should be transformed via the `RuntimeEnumExtender`
> - Second, add a `getExtensionInfo` method which returns `ExtensionInfo.nonExtended(TheEnum.class)`

Reviewer: ```suggestion
- Make the enum implement `IExtensibleEnum` to mark that this enum should be transformed via the `RuntimeEnumExtender`.
- Add a `getExtensionInfo` method that returns `ExtensionInfo.nonExtended(TheEnum.class)`.
```

Further action is required depending on specific details about the enum:

- If the enum has an int ID parameter which should match the entry's ordinal, then the enum should be annotated with `@NumberedEnum` with the ID's parameter index as the annotation's value if it's not the first parameter

> - If the enum has a String name parameter which is used for serialization and should therefore be namespaced, then the enum should be annotated with `@NamedEnum` with the name's parameter index as the annotation's value if it's not the first parameter

Reviewer: I am confused as to what the annotation parameters are supposed to be, needs improved wording.

Author: The annotation parameter is the relevant constructor parameter's index. In most cases that would be 0 (the annotation's default value), except when the given constructor parameter is not the first one (i.e. when an enum has an ID and a name parameter as the first and second parameter respectively, then index 1 needs to be specified in the annotation for the latter). Not quite sure at the moment how to word that better.

All enums that can have new entries implement the `IExtensibleEnum` interface. This interface acts as a marker to allow the `RuntimeEnumExtender` launch plugin service to know what enums should be transformed.

:::warning
You should **not** be implementing this interface on your own enums. Use maps or registries instead depending on your usecase.  
Enums which are not patched to implement the interface cannot have the interface added to them via mixins or coremods due to the order the transformers run in.
:::

### Creating an Enum Entry

To create new enum entries, a JSON file needs to be created and referenced in the `neoforge.mods.toml` with the `enumExtensions` entry of a `[[mods]]` block. The specified path must be relative to the `resources` directory:
```toml
# In neoforge.mods.toml:
[[mods]]
## The file is relative to the output directory of the resources, or the root path inside the jar when compiled
## The 'resources' directory represents the root output directory of the resources

> enumExtensions="META-INF/enumextensions.json"

Reviewer 2: I think this also needs to be added to modfiles.md as a new option as well. That can probably link to this page.

      // The parameters to be used, provided as a reference to a method in the given class
      "parameters": {
        "class": "example/examplemod/MyEnumParams",
        "field": "getTestDamageEffectsParameter"
      }
    },
    {
      "enum": "net/minecraft/world/item/ItemDisplayContext",
      "name": "EXAMPLEMOD_STANDING",
      "constructor": "(ILjava/lang/String;Ljava/lang/String;)V",
      // Constant parameters provided directly.
      "parameters": [ -1, "examplemod:standing", null ]
    }
  ]
}
```

```java
public class MyEnumParams {
    public static final EnumProxy<Rarity> CUSTOM_RARITY_ENUM_PROXY = new EnumProxy<>(
            Rarity.class, -1, "examplemod:custom", (UnaryOperator<Style>) style -> style.withItalic(true)
    );
    
    public static Object getTestDamageEffectsParameter(int idx, Class<?> type) {
        return type.cast(switch (idx) {
            case 0 -> "examplemod:test";
            case 1 -> (Supplier<SoundEvent>) () -> SoundEvents.DONKEY_ANGRY;
            default -> throw new IllegalArgumentException("Unexpected parameter index: " + idx);
        });
    }
}
```

#### Constructor

> The constructor must be specified as a method descriptor and must only contain the parameters visible in the source code, omitting the hidden constant name and ordinal parameters.

Reviewer 2: Probably a link to the JVMS page similar to Access transformers on how to write out the definition.

#### Parameters

The parameters can be specified in three ways with limitations depending on the parameter types:

- Inline in the JSON file as an array of constants (only allowed for primitive values, Strings and for passing null to any reference type)
- As a reference to a field of type `EnumProxy<TheEnum>` in a class from the mod (see `EnumProxy` example above)
- As a reference to a method returning `Object`, where the return value is the parameter value to use. The method must have exactly two parameters of type `int` (index of the parameter) and `Class<?>` (expected type of the parameter)

>   - The `Class<?>` object should be used to cast (`Class#cast()`) the return value in order to keep `ClassCastException`s in mod code.

Reviewer 2: I would order these how you order them in the above example. Or, it would probably be better to construct a separate example representing how each is used, because this feels a bit too brief.

:::warning
The fields and/or methods used as sources for parameter values should be in a separate class to avoid unintentionally loading mod classes too early.
:::

Certain parameters have additional rules:

> - If the parameter is an int ID parameter related to a `@IndexedEnum` annotation on the enum, then it is ignored and replaced by the entry's ordinal. If said parameter is specified inline in the JSON, then it must be specified as `-1`.

Reviewer 2: Does this throw an error if not -1? And how does this work if the index enum is not the first parameter (I don't think this happens in vanilla, but I'm unsure whether ignore means only if its the first parameter).

> - If the parameter is a String name parameter related to a `@NamedEnum` annotation on the enum, then it must be prefixed by the mod ID in the `namespace:path` format known from `ResourceLocation`s.

Reviewer 2: Does this throw an error if not prefixed?

## Contributing to NeoForge

To add a new extensible enum to NeoForge, there are at least two required things to do:

- Make the enum implement `IExtensibleEnum` to mark that this enum should be transformed via the `RuntimeEnumExtender`.
- Add a `getExtensionInfo` method that returns `ExtensionInfo.nonExtended(TheEnum.class)`.

Further action is required depending on specific details about the enum:

- If the enum has an int ID parameter which should match the entry's ordinal, then the enum should be annotated with `@NumberedEnum` with the ID's parameter index as the annotation's value if it's not the first parameter
- If the enum has a String name parameter which is used for serialization and should therefore be namespaced, then the enum should be annotated with `@NamedEnum` with the name's parameter index as the annotation's value if it's not the first parameter
- If the enum is sent over the network, then it should be annotated with `@NetworkedEnum` with the annotation's parameter specifying in which direction the values may be sent (clientbound, serverbound or bidirectional)

>   - TODO: add to network check if necessary

Reviewer 2: Ah yes, a lovely todo

> - If the enum has constructors which are not usable by mods (i.e. because they require registry objects on an enum that may be initialized before modded registration runs), then they should be annotated with `@ReservedConstructor`

Reviewer 2: This should also be explained above when creating an extensible enum since this means that the constructor cannot be selected.

    {
      // The enum class the entry should be added to
      "enum": "net/minecraft/world/item/ItemDisplayContext",
      // The field name of the new entry, must be prefixed with the mod ID
      "name": "EXAMPLEMOD_STANDING",
      // The constructor to be used
      "constructor": "(ILjava/lang/String;Ljava/lang/String;)V",
      // Constant parameters provided directly.
      "parameters": [ -1, "examplemod:standing", null ]
    },
    {
      "enum": "net/minecraft/world/item/Rarity",
      "name": "EXAMPLEMOD_CUSTOM",
      "constructor": "(ILjava/lang/String;Ljava/util/function/UnaryOperator;)V",
      // The parameters to be used, provided as a reference to an EnumProxy<Rarity> field in the given class
      "parameters": {
        "class": "example/examplemod/MyEnumParams",
        "field": "CUSTOM_RARITY_ENUM_PROXY"
      }
    },
    {
      "enum": "net/minecraft/world/damagesource/DamageEffects",
      "name": "EXAMPLEMOD_TEST",
      "constructor": "(Ljava/lang/String;Ljava/util/function/Supplier;)V",
      // The parameters to be used, provided as a reference to a method in the given class
      "parameters": {
        "class": "example/examplemod/MyEnumParams",

>         "field": "getTestDamageEffectsParameter"

Reviewer 3: Without looking at the system, I'm going to assume this is incorrect because the key is "field", and this is referencing a method.

Author: Indeed, copy-paste screwed me over once again

>       }

Reviewer 2: Four spaces per indentation.

            case 0 -> "examplemod:test";
            case 1 -> (Supplier<SoundEvent>) () -> SoundEvents.DONKEY_ANGRY;
            default -> throw new IllegalArgumentException("Unexpected parameter index: " + idx);
        });
    }
}
```

#### Constructor

The constructor must be specified as a [method descriptor][jvmdescriptors] and must only contain the parameters visible in the source code, omitting the hidden constant name and ordinal parameters.  
If a constructor is marked with the `@ReservedConstructor` annotation, then it cannot be used for modded enum constants.

#### Parameters

The parameters can be specified in three ways with limitations depending on the parameter types:

- Inline in the JSON file as an array of constants (only allowed for primitive values, Strings and for passing null to any reference type)
- As a reference to a field of type `EnumProxy<TheEnum>` in a class from the mod (see `EnumProxy` example above)
- As a reference to a method returning `Object`, where the return value is the parameter value to use. The method must have exactly two parameters of type `int` (index of the parameter) and `Class<?>` (expected type of the parameter)

>   - The `Class<?>` object should be used to cast (`Class#cast()`) the return value in order to keep `ClassCastException`s in mod code.

Reviewer 2: So, if `Class#cast` should be used to cast the return value, why isn't that done within the example?

Author: The example does use it, the switch is wrapped in the `Class#cast()` call. I can extract it to a separate line if you prefer that.

Reviewer 2: Ah, my bad. I simply missed it because it was wrapping around the switch statement.
